File: docs/concepts/chapters/leblanc/r3-method/seam-repaint.mjs
#!/usr/bin/env node
/**
 * Art method r3 pilot, P3/P4 (FFX-2 only, chapter 6 Leblanc): the seam-band-only repaint.
 * A pilot copy of tools/gen/lora-repaint.mjs (the shared tool is untouched), changed as
 * METHOD-CHECK step 2.5 asks: Animagine XL 4.0 Opt + leblanc-x2-r2 at 0.8, IP-Adapter plus
 * (vit-h) on the idle square-padded + the idle head crop at 0.3 (ease in, 0.2 to 0.6, K+V,
 * concat: round 2's settings), VAEEncode + SetLatentNoiseMask, low denoise, and words that
 * name only what the band shows. The init and mask are the crop seam.py prep wrote
 * (region plus 25 % margin, 1024 short side).
 *
 *   node docs/concepts/chapters/leblanc/r3-method/seam-repaint.mjs --dir p3 --region obi,choker
 *        [--denoise 0.25,0.35] [--seeds 4] [--seed 83000] [--gate 3] [--tag c]
 *
 * Writes <dir>/<region>.<tag><i>.png and .json under D:/Tools/pyrefly-lora/leblanc/r3/. One
 * prompt at a time behind an empty shared queue (the first prompt waits for the queue to be
 * empty for --gate minutes); ComfyUI is never restarted from here. Every prompt id and its
 * execution time from the history API is appended to gpu-ledger.json; the run stops at the
 * 60-minute pilot cap and on an all-black frame (hard rule 12).
 */
import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'node:fs';
import { join, resolve, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { BASE_NEGATIVE, stageImage } from '../../../../../tools/gen/comfy.mjs';

const HERE = dirname(fileURLToPath(import.meta.url));
const REPO = resolve(HERE, '..', '..', '..', '..', '..');
const SCR = 'D:/Tools/pyrefly-lora/leblanc/r3';
const LEDGER = join(SCR, 'gpu-ledger.json');
const COMFY = process.env.COMFY_URL || 'http://127.0.0.1:8188';
const CKPT = 'animagine-xl-4.0-opt.safetensors';
const LORA = 'leblanc-x2-r2.safetensors';
const IPADAPTER = 'ip-adapter-plus_sdxl_vit-h.safetensors';
const CLIPVISION = 'CLIP-ViT-H-14-laion2B-s32B-b79K.safetensors';
const REFS = ['docs/concepts/chapters/leblanc/pilot2/refs/idle-square.png', 'docs/concepts/chapters/leblanc/pilot2/refs/idle-head.png'];
const CAP_MIN = 60;
const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

/** Words that name only what each band shows (C2: colour words bind to the wrong noun, so none on the parts). */
export const WORDS = {
  obi: 'leblancX2, 1girl, close-up, midriff, white dress, obi, tassel, black lining, purple kimono, thigh, pale skin',
  choker: 'leblancX2, 1girl, close-up, neck, choker, pale skin, blonde hair, white dress, halterneck',
  hurt: 'leblancX2, 1girl, close-up, purple kimono, white dress, pale skin, blonde hair, bob cut',
  wince: 'leblancX2, 1girl, close-up, face, closed eyes, wince, pain, blonde hair, bob cut, pale skin',
};
const EXTRA_NEG = 'multiple views, 2girls, chibi, sketch, monochrome, 3d, realistic, gold obi, navy obi, two tassels, extra tassel, second tassel, text, watermark';

function parse() {
  const a = {};
  const v = process.argv.slice(2);
  for (let i = 0; i < v.length; i++) if (v[i].startsWith('--')) a[v[i].slice(2)] = v[i + 1] && !v[i + 1].startsWith('--') ? v[++i] : true;
  return a;
}

export function ledgerMinutes() {
  if (!existsSync(LEDGER)) return 0;
  return JSON.parse(readFileSync(LEDGER, 'utf8')).reduce((s, e) => s + (e.execMs || 0), 0) / 60000;
}

function addLedger(e) {
  const all = existsSync(LEDGER) ? JSON.parse(readFileSync(LEDGER, 'utf8')) : [];
  all.push(e);
  writeFileSync(LEDGER, JSON.stringify(all, null, 1));
}

async function queueEmpty() {
  try {
    const q = await (await fetch(`${COMFY}/queue`)).json();
    return !q.queue_running.length && !q.queue_pending.length;
  } catch {
    return false;
  }
}

/** P0: the queue empty for `minutes` consecutive minutes. */
async function gate(minutes) {
  let since = null;
  for (;;) {
    if (await queueEmpty()) {
      since ??= Date.now();
      if (Date.now() - since >= minutes * 60_000) return;
    } else since = null;
    await sleep(10_000);
  }
}

export function graph({ init, mask, refA, refB, seed, denoise, words, negative, strength = 0.8, refWeight = 0.3, prefix }) {
  return {
    4: { class_type: 'CheckpointLoaderSimple', inputs: { ckpt_name: CKPT } },
    10: { class_type: 'LoraLoader', inputs: { model: ['4', 0], clip: ['4', 1], lora_name: LORA, strength_model: strength, strength_clip: strength } },
    6: { class_type: 'CLIPTextEncode', inputs: { text: words, clip: ['10', 1] } },
    7: { class_type: 'CLIPTextEncode', inputs: { text: negative, clip: ['10', 1] } },
    20: { class_type: 'LoadImage', inputs: { image: refA, upload: 'image' } },
    24: { class_type: 'LoadImage', inputs: { image: refB, upload: 'image' } },
    25: { class_type: 'ImageBatch', inputs: { image1: ['20', 0], image2: ['24', 0] } },
    21: { class_type: 'IPAdapterModelLoader', inputs: { ipadapter_file: IPADAPTER } },
    22: { class_type: 'CLIPVisionLoader', inputs: { clip_name: CLIPVISION } },
    23: {
      class_type: 'IPAdapterAdvanced',
      inputs: {
        model: ['10', 0], ipadapter: ['21', 0], image: ['25', 0], weight: refWeight, weight_type: 'ease in',
        combine_embeds: 'concat', start_at: 0.2, end_at: 0.6, embeds_scaling: 'K+V', clip_vision: ['22', 0],
      },
    },
    40: { class_type: 'LoadImage', inputs: { image: init, upload: 'image' } },
    41: { class_type: 'VAEEncode', inputs: { pixels: ['40', 0], vae: ['4', 2] } },
    42: { class_type: 'LoadImage', inputs: { image: mask, upload: 'image' } },
    43: { class_type: 'ImageToMask', inputs: { image: ['42', 0], channel: 'red' } },
    44: { class_type: 'SetLatentNoiseMask', inputs: { samples: ['41', 0], mask: ['43', 0] } },
    3: {
      class_type: 'KSampler',
      inputs: { seed, steps: 30, cfg: 6, sampler_name: 'euler_ancestral', scheduler: 'normal', denoise, model: ['23', 0], positive: ['6', 0], negative: ['7', 0], latent_image: ['44', 0] },
    },
    8: { class_type: 'VAEDecode', inputs: { samples: ['3', 0], vae: ['4', 2] } },
    9: { class_type: 'SaveImage', inputs: { filename_prefix: prefix, images: ['8', 0] } },
  };
}

export async function run(workflow, outPath, note) {
  if (ledgerMinutes() >= CAP_MIN) throw new Error(`pilot GPU cap ${CAP_MIN} min reached (${ledgerMinutes().toFixed(1)}); stop and report`);
  let noted = 0;
  while (!(await queueEmpty())) {
    if (Date.now() - noted > 60_000) { console.error('[seam] queue busy; waiting'); noted = Date.now(); }
    await sleep(5000);
  }
  const res = await fetch(`${COMFY}/prompt`, { method: 'POST', headers: { 'content-type': 'application/json' }, body: JSON.stringify({ prompt: workflow, client_id: 'pyrefly-r3-seam' }) });
  const body = await res.json();
  if (!res.ok || body.error) throw new Error(`ComfyUI rejected: ${JSON.stringify(body).slice(0, 1500)}`);
  const id = body.prompt_id;
  for (;;) {
    const e = (await (await fetch(`${COMFY}/history/${id}`)).json())[id];
    if (e?.status?.status_str === 'error') throw new Error(`failed: ${JSON.stringify(e.status.messages).slice(0, 1500)}`);
    const img = e && Object.values(e.outputs || {}).flatMap((o) => o.images || [])[0];
    if (img && e.status?.completed) {
      const q = new URLSearchParams({ filename: img.filename, subfolder: img.subfolder || '', type: img.type || 'output' });
      writeFileSync(outPath, Buffer.from(await (await fetch(`${COMFY}/view?${q}`)).arrayBuffer()));
      const msgs = Object.fromEntries((e.status.messages || []).map(([k, v]) => [k, v.timestamp]));
      const execMs = (msgs.execution_success ?? 0) - (msgs.execution_start ?? 0);
      addLedger({ id, out: outPath, execMs, note, at: new Date().toISOString() });
      return execMs / 1000;
    }
    await sleep(1000);
  }
}

async function main() {
  const a = parse();
  const dir = join(SCR, a.dir || 'p3');
  const regions = String(a.region || 'obi,choker').split(',');
  const denoise = String(a.denoise ?? '0.25,0.35').split(',').map(Number);
  const seeds = Number(a.seeds ?? 4);
  const seed0 = Number(a.seed ?? 83000);
  const tag = a.tag || 'c';
  const negative = `${BASE_NEGATIVE}, ${EXTRA_NEG}${a.neg ? `, ${a.neg}` : ''}`;
  mkdirSync(dir, { recursive: true });
  if (a.gate !== '0') {
    console.log(`[seam] P0 gate: waiting for the queue to be empty for ${a.gate ?? 3} min`);
    await gate(Number(a.gate ?? 3));
  }
  const refA = stageImage(join(REPO, REFS[0]));
  const refB = stageImage(join(REPO, REFS[1]));
  for (const region of regions) {
    const words = a.words || WORDS[a.wordset || region] || WORDS.hurt;
    const init = stageImage(join(dir, `${region}.init.png`));
    const mask = stageImage(join(dir, `${region}.mask.png`));
    for (let i = 0; i < seeds; i++) {
      const d = denoise[i % denoise.length];
      const seed = seed0 + i;
      const out = join(dir, `${region}.${tag}${i + 1}`);
      if (existsSync(`${out}.png`)) { console.log(`[seam] ${out}.png exists; skip`); continue; }
      const g = graph({ init, mask, refA, refB, seed, denoise: d, words, negative, prefix: 'pyrefly/r3-seam' });
      const secs = await run(g, `${out}.png`, `${region} ${tag}${i + 1} d${d} s${seed}`);
      writeFileSync(`${out}.json`, JSON.stringify({ region, init: `${region}.init.png`, mask: `${region}.mask.png`, seed, denoise: d, words, negative, lora: { file: LORA, strength: 0.8 },
        ipadapter: { file: IPADAPTER, weight: 0.3, type: 'ease in', start: 0.2, end: 0.6, scaling: 'K+V', combine: 'concat', images: REFS }, model: CKPT, steps: 30, cfg: 6,
        sampler: 'euler_ancestral', scheduler: 'normal', execSeconds: secs }, null, 1));
      console.log(`[seam] ${out}.png d${d} seed ${seed}: ${secs.toFixed(1)} s (ledger ${ledgerMinutes().toFixed(2)} min)`);
    }
  }
}

const isMain = process.argv[1] && resolve(process.argv[1]) === fileURLToPath(import.meta.url);
if (isMain) main().catch((e) => { console.error(`[seam] FAILED: ${e.message}`); process.exit(1); });

File: docs/concepts/chapters/evrae/r3/scripts/seam_repaint.py
"""Seam-band-only repaint (FFX only), METHOD-CHECK step 2.5: crop the region, upscale to a
1024-ish canvas on flat grey, img2img Animagine XL 4.0 Opt with SetLatentNoiseMask over the
band mask only, low denoise; Lanczos back; composite inside the band only (feathered).
One prompt at a time; waits for the shared queue to be empty for 3 minutes first.
usage: seam_repaint.py <rgba.png> <band_mask.png (L, white = repaint)> <x0,y0,x1,y1> <denoise> <seed> <out.png> "<tags>" """
import sys, json, time, uuid, urllib.request, urllib.parse, io, os
import numpy as np, cv2
COMFY = 'http://127.0.0.1:8188'
src, maskp, box, den, seed, out, tags = sys.argv[1:8]
x0, y0, x1, y1 = map(int, box.split(',')); den = float(den); seed = int(seed)
im = cv2.imread(src, -1).astype(np.float32) / 255
band = cv2.imread(maskp, 0).astype(np.float32) / 255
crop = im[y0:y1, x0:x1]; bc = band[y0:y1, x0:x1]
s = 1024 / min(crop.shape[:2]); W = int(round(crop.shape[1] * s / 8) * 8); H = int(round(crop.shape[0] * s / 8) * 8)
a = crop[..., 3:4]; rgb = crop[..., :3] * a + 0.5 * (1 - a)
big = cv2.resize(rgb, (W, H), interpolation=cv2.INTER_LANCZOS4)
bm = cv2.resize(bc, (W, H), interpolation=cv2.INTER_LINEAR)
def get(path):
    return json.loads(urllib.request.urlopen(COMFY + path, timeout=60).read())
def upload(arr, name):
    ok, buf = cv2.imencode('.png', arr)
    b = uuid.uuid4().hex
    body = (f'--{b}\r\nContent-Disposition: form-data; name="image"; filename="{name}"\r\nContent-Type: image/png\r\n\r\n').encode() + buf.tobytes() + f'\r\n--{b}\r\nContent-Disposition: form-data; name="overwrite"\r\n\r\ntrue\r\n--{b}--\r\n'.encode()
    r = urllib.request.Request(COMFY + '/upload/image', data=body, headers={'Content-Type': f'multipart/form-data; boundary={b}'})
    return json.loads(urllib.request.urlopen(r, timeout=60).read())['name']
# the shared-GPU rule: queue empty for 3 minutes
calm = None
while True:
    q = get('/queue')
    if q['queue_running'] or q['queue_pending']: calm = None
    elif calm is None: calm = time.time()
    if calm and time.time() - calm >= 180: break
    time.sleep(10)
iname = upload(np.clip(big * 255, 0, 255).astype(np.uint8), f'evrae-r3-{seed}-init.png')
mname = upload(np.clip(np.dstack([bm] * 3) * 255, 0, 255).astype(np.uint8), f'evrae-r3-{seed}-mask.png')
neg = 'lowres, bad anatomy, text, error, worst quality, low quality, low score, bad score, average score, jpeg artifacts, signature, watermark, username, blurry, artist name, human, person, flat color, vector art, gradient, sticker'
pos = 'no humans, evrae \(ff10\), final fantasy x, monster, eastern dragon, sea serpent, ' + tags + ', painterly, official art, soft shading, detailed scales, masterpiece, high score, great score, absurdres'
wf = {
 '1': {'class_type': 'CheckpointLoaderSimple', 'inputs': {'ckpt_name': 'animagine-xl-4.0-opt.safetensors'}},
 '2': {'class_type': 'LoadImage', 'inputs': {'image': iname}},
 '3': {'class_type': 'LoadImageMask', 'inputs': {'image': mname, 'channel': 'red'}},
 '4': {'class_type': 'VAEEncode', 'inputs': {'pixels': ['2', 0], 'vae': ['1', 2]}},
 '5': {'class_type': 'SetLatentNoiseMask', 'inputs': {'samples': ['4', 0], 'mask': ['3', 0]}},
 '6': {'class_type': 'CLIPTextEncode', 'inputs': {'text': pos, 'clip': ['1', 1]}},
 '7': {'class_type': 'CLIPTextEncode', 'inputs': {'text': neg, 'clip': ['1', 1]}},
 '8': {'class_type': 'KSampler', 'inputs': {'model': ['1', 0], 'positive': ['6', 0], 'negative': ['7', 0], 'latent_image': ['5', 0], 'seed': seed, 'steps': 28, 'cfg': 5.5, 'sampler_name': 'euler_ancestral', 'scheduler': 'normal', 'denoise': den}},
 '9': {'class_type': 'VAEDecode', 'inputs': {'samples': ['8', 0], 'vae': ['1', 2]}},
 '10': {'class_type': 'SaveImage', 'inputs': {'images': ['9', 0], 'filename_prefix': 'evrae-r3/seam'}},
}
r = urllib.request.Request(COMFY + '/prompt', data=json.dumps({'prompt': wf, 'client_id': 'evrae-r3'}).encode(), headers={'Content-Type': 'application/json'})
pid = json.loads(urllib.request.urlopen(r, timeout=60).read())['prompt_id']
while True:
    h = get('/history/' + pid)
    if pid in h and h[pid].get('outputs'):
        break
    if pid in h and h[pid].get('status', {}).get('status_str') == 'error': raise SystemExit('comfy error ' + json.dumps(h[pid]['status']))
    time.sleep(3)
e = h[pid]
img = [i for o in e['outputs'].values() for i in o.get('images', [])][0]
q = urllib.parse.urlencode({'filename': img['filename'], 'subfolder': img.get('subfolder', ''), 'type': img.get('type', 'output')})
gen = cv2.imdecode(np.frombuffer(urllib.request.urlopen(COMFY + '/view?' + q).read(), np.uint8), cv2.IMREAD_COLOR).astype(np.float32) / 255
msgs = e.get('status', {}).get('messages', [])
ts = [m[1].get('timestamp') for m in msgs if m[0] in ('execution_start', 'execution_success')]
secs = (ts[-1] - ts[0]) / 1000 if len(ts) >= 2 else None
small = cv2.resize(gen, (x1 - x0, y1 - y0), interpolation=cv2.INTER_AREA)
k = cv2.GaussianBlur(bc, (0, 0), 1.2)[..., None]
res = im.copy()
res[y0:y1, x0:x1, :3] = crop[..., :3] * (1 - k) + small * k
save = np.clip(res * 255 + 0.5, 0, 255).astype(np.uint8)
cv2.imwrite(out, save)
if gen.mean() < 0.02: print('ALL-BLACK FRAME: bad GPU state, stop'); sys.exit(3)
json.dump({'prompt_id': pid, 'seconds': secs, 'seed': seed, 'denoise': den, 'box': [x0, y0, x1, y1], 'canvas': [W, H], 'tags': tags}, open(out.replace('.png', '.json'), 'w'), indent=1)
print(json.dumps({'prompt_id': pid, 'seconds': secs, 'mean': float(gen.mean())}))
